include/device.h

```diff
@@ -329,6 +345,29 @@ size_t z_device_get_all_static(const struct device * *devices);
  */
 bool z_device_ready(const struct device *dev);
 
+/** @brief Verify that a device is ready for use.
+ *
+ * Returns the provided device pointer if the device is known to be in a state
+ * where it can be used with its standard API.  If the device cannot be used a
+ * null pointer is returned.
+ *
+ * This can be used with device pointers captured from DEVICE_DT_GET(), which
+ * does not include the readiness checks of device_get_binding().  At minimum
+ * this means that the device has been successfully initialized, but it may
+ * take on further conditions (e.g. is not powered down).
+ *
+ * @param dev pointer to the device in question.
+ *
+ * @return @p dev if the device is ready for use, or a null pointer if
+ * it is not.
+ */
+__syscall const struct device *device_if_ready(const struct device *dev);
```

Comment:
not a great API name, maybe device_is_ready() or device_ready_get()

Reply:
`device_is_ready()` suggests a boolean return, which is why it's `device_if_ready()`.

I guess we could do `device_ready_get()`.  Other opinions?

Reply:
how about `device_is_ready` or `is_device_ready`

Reply:
Or should we just go ahead and implement `DEVICE_DT_GET_BINDING()` and have it do the ready check and drop `device_if_ready`?

Reply:
No, we need `device_if_ready()` for drivers where the device pointer is stored in flash at build time and needs to be validated at runtime

The original idiom plan was:

```
const struct device *bus = device_if_ready(cfg->bus);

if (bus == NULL) {
  return -ENOENT;
}
... /* use bus */
```

and

```
const struct device *bus = device_if_ready(DEVICE_DT_GET(BUS_NODE));

if (bus == NULL) {
  return -ENOENT;
}
... /* use bus */
```

If we go with an is-ready it'll be:

```
if (!device_is_ready(cfg->bus)) {
  return -ENOENT;
}
... /* use cfg->bus */
```

and

```
const struct device *bus = DEVICE_DT_GET(BUS_NODE);

if (!device_is_ready(bus)) {
  return -ENOENT;
}
... /* use bus */
```

I preferred the first because it's structurally consistent, but if people really don't want `device_if_ready()` I'll go with the second.

Reply:
> I guess we could do `device_ready_get()`.

yeah, works for me.

Reply:
I personally prefer device_if_ready but don't have a strong opinion.

Reply:
I would vote for `device_get_ready()` instead

Reply:
I don't want `get` because `get/put` are used in power management for reference counting access.  It's now `device_is_ready()` and I've updated shell use of the internal API to use the public version.

Reply:
I really dislike `device_is_ready()`. I suggest we instead use `device_ready()` to avoid the inconsistency between the predicate and not returning a bool. I will add this to the API meeting to see if we can close this today.